From: Simon McVittie <smcv@debian.org>
Date: Sat, 13 Feb 2021 18:26:24 +0000
Subject: reftest_compare_surfaces: Report how much the images differ

In unattended/non-interactive/autobuilder environments where the images
are not trivially accessible, this provides a way to distinguish between
totally different rendering and more subtle issues.

Signed-off-by: Simon McVittie <smcv@debian.org>
Forwarded: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/3195
Applied-upstream: no, upstream want reftests to be a strict pass/fail with identical results required
---
 testsuite/gsk/compare-render.c       | 33 +++++++++++++++++++++++++++------
 testsuite/reftests/gtk-reftest.c     |  9 ++++++++-
 testsuite/reftests/reftest-compare.c | 28 +++++++++++++++++++++++++---
 testsuite/reftests/reftest-compare.h |  6 +++++-
 4 files changed, 65 insertions(+), 11 deletions(-)

Index: gtk4-4.14.1+ds/testsuite/gsk/compare-render.c
===================================================================
--- gtk4-4.14.1+ds.orig/testsuite/gsk/compare-render.c	2024-03-20 21:24:00.036733677 +0100
+++ gtk4-4.14.1+ds/testsuite/gsk/compare-render.c	2024-03-20 21:24:00.036733677 +0100
@@ -318,6 +318,9 @@
   GError *error = NULL;
   GOptionContext *context;
   GdkTexture *diff_texture = NULL;
+  guint max_diff = 0;
+  guint pixels_changed = 0;
+  guint pixels = 0;
 
   (g_test_init) (&argc, &argv, NULL);
 
@@ -377,9 +380,12 @@
         }
 
       /* Now compare the two */
-      diff_texture = reftest_compare_textures (rendered_texture, reference_texture);
+      diff_texture = reftest_compare_textures (rendered_texture, reference_texture,
+                                               &max_diff, &pixels_changed, &pixels);
       if (diff_texture)
         {
+          g_print ("%u (out of %u) pixels differ from reference by up to %u levels\n",
+                   pixels_changed, pixels, max_diff); 
           save_image (diff_texture, node_file, ".diff.png");
           success = FALSE;
         }
@@ -412,10 +418,13 @@
 
       save_image (reference_texture, node_file, "-flipped.ref.png");
 
-      diff_texture = reftest_compare_textures (rendered_texture, reference_texture);
+      diff_texture = reftest_compare_textures (rendered_texture, reference_texture,
+                                               &max_diff, &pixels_changed, &pixels);
 
       if (diff_texture)
         {
+          g_print ("%u (out of %u) pixels differ from reference by up to %u levels\n",
+                   pixels_changed, pixels, max_diff);
           save_image (diff_texture, node_file, "-flipped.diff.png");
           success = FALSE;
         }
@@ -478,10 +487,13 @@
 
       save_image (reference_texture, node_file, "-repeated.ref.png");
 
-      diff_texture = reftest_compare_textures (rendered_texture, reference_texture);
+      diff_texture = reftest_compare_textures (rendered_texture, reference_texture,
+                                               &max_diff, &pixels_changed, &pixels);
 
       if (diff_texture)
         {
+          g_print ("%u (out of %u) pixels differ from reference by up to %u levels\n",
+                   pixels_changed, pixels, max_diff);
           save_image (diff_texture, node_file, "-repeated.diff.png");
           success = FALSE;
         }
@@ -515,10 +527,13 @@
 
       save_image (reference_texture, node_file, "-rotated.ref.png");
 
-      diff_texture = reftest_compare_textures (rendered_texture, reference_texture);
+      diff_texture = reftest_compare_textures (rendered_texture, reference_texture,
+                                               &max_diff, &pixels_changed, &pixels);
 
       if (diff_texture)
         {
+          g_print ("%u (out of %u) pixels differ from reference by up to %u levels\n",
+                   pixels_changed, pixels, max_diff);
           save_image (diff_texture, node_file, "-rotated.diff.png");
           success = FALSE;
         }
@@ -572,10 +587,13 @@
 
       save_image (reference_texture, node_file, "-masked.ref.png");
 
-      diff_texture = reftest_compare_textures (rendered_texture, reference_texture);
+      diff_texture = reftest_compare_textures (rendered_texture, reference_texture,
+                                               &max_diff, &pixels_changed, &pixels);
 
       if (diff_texture)
         {
+          g_print ("%u (out of %u) pixels differ from reference by up to %u levels\n",
+                   pixels_changed, pixels, max_diff);
           save_image (diff_texture, node_file, "-masked.diff.png");
           success = FALSE;
         }
@@ -613,10 +631,13 @@
       g_assert_nonnull (rendered_texture);
       g_assert_nonnull (rendered_texture2);
 
-      diff_texture = reftest_compare_textures (rendered_texture, rendered_texture2);
+      diff_texture = reftest_compare_textures (rendered_texture, reference_texture,
+                                               &max_diff, &pixels_changed, &pixels);
 
       if (diff_texture)
         {
+          g_print ("%u (out of %u) pixels differ from reference by up to %u levels\n",
+                   pixels_changed, pixels, max_diff);
           save_image (diff_texture, node_file, "-replayed.diff.png");
           success = FALSE;
         }
@@ -668,7 +689,8 @@
 
       save_image (reference_texture, node_file, "-clipped.ref.png");
 
-      diff_texture = reftest_compare_textures (rendered_texture, reference_texture);
+      diff_texture = reftest_compare_textures (rendered_texture, reference_texture,
+                                               &max_diff, &pixels_changed, &pixels);
 
       if (diff_texture)
         {
Index: gtk4-4.14.1+ds/testsuite/reftests/gtk-reftest.c
===================================================================
--- gtk4-4.14.1+ds.orig/testsuite/reftests/gtk-reftest.c	2024-03-20 21:24:00.036733677 +0100
+++ gtk4-4.14.1+ds/testsuite/reftests/gtk-reftest.c	2024-03-20 21:24:00.036733677 +0100
@@ -341,6 +341,9 @@
   char *ui_file, *reference_file;
   GdkTexture *ui_image, *reference_image, *diff_image;
   GtkStyleProvider *provider;
+  guint max_diff = 0;
+  guint pixels_changed = 0;
+  guint pixels = 0;
 
   ui_file = g_file_get_path (file);
 
@@ -372,12 +375,16 @@
   if (reference_image == NULL)
     reference_image = gdk_memory_texture_new (1, 1, GDK_MEMORY_DEFAULT, g_bytes_new ((guchar[4]) {0, 0, 0, 0}, 4), 4);
 
-  diff_image = reftest_compare_textures (ui_image, reference_image);
+  diff_image = reftest_compare_textures (ui_image, reference_image,
+                                         &max_diff, &pixels_changed, &pixels);
 
   save_image (ui_image, ui_file, ".out.png");
   save_image (reference_image, ui_file, ".ref.png");
+
   if (diff_image)
     {
+      g_test_message ("%u (out of %u) pixels differ from reference by up to %u levels",
+                      pixels_changed, pixels, max_diff);
       save_node (g_object_get_data (G_OBJECT (ui_image), "source-render-node"), ui_file, ".out.node");
       save_node (g_object_get_data (G_OBJECT (reference_image), "source-render-node"), ui_file, ".ref.node");
       save_image (diff_image, ui_file, ".diff.png");
Index: gtk4-4.14.1+ds/testsuite/reftests/reftest-compare.c
===================================================================
--- gtk4-4.14.1+ds.orig/testsuite/reftests/reftest-compare.c	2024-03-20 21:24:00.036733677 +0100
+++ gtk4-4.14.1+ds/testsuite/reftests/reftest-compare.c	2024-03-20 21:24:00.036733677 +0100
@@ -35,12 +35,16 @@
         	  const guchar *buf_b,
                   int           stride_b,
         	  int		width,
-        	  int		height)
+		  int		height,
+                  guint        *max_diff_out,
+                  guint        *pixels_changed_out)
 {
   int x, y;
   guchar *buf_diff = NULL;
   int stride_diff = 0;
   GdkTexture *diff = NULL;
+  guint max_diff = 0;
+  guint pixels_changed = 0;
 
   for (y = 0; y < height; y++)
     {
@@ -84,6 +88,10 @@
               guint channel_diff;
 
               channel_diff = ABS (value_a - value_b);
+
+              if (channel_diff > max_diff)
+                max_diff = channel_diff;
+
               channel_diff *= 4;  /* emphasize */
               if (channel_diff)
                 channel_diff += 128; /* make sure it's visible */
@@ -92,6 +100,8 @@
               diff_pixel |= channel_diff << (channel * 8);
             }
 
+          pixels_changed++;
+
           if ((diff_pixel & 0x00ffffff) == 0)
             {
               /* alpha only difference, convert to luminance */
@@ -105,12 +115,21 @@
       }
   }
 
+  if (max_diff_out != NULL)
+    *max_diff_out = max_diff;
+
+  if (pixels_changed_out != NULL)
+    *pixels_changed_out = pixels_changed;
+
   return diff;
 }
 
 GdkTexture *
 reftest_compare_textures (GdkTexture *texture1,
-                          GdkTexture *texture2)
+                          GdkTexture *texture2,
+                          guint      *max_diff_out,
+                          guint      *pixels_changed_out,
+                          guint      *pixels_out)
 {
   int w, h;
   guchar *data1, *data2;
@@ -126,10 +145,13 @@
 
   diff = buffer_diff_core (data1, w * 4,
                            data2, w * 4,
-                           w, h);
+                           w, h, max_diff_out, pixels_changed_out);
 
   g_free (data1);
   g_free (data2);
 
+  if (pixels_out != NULL)
+    *pixels_out = w * h;
+
   return diff;
 }
Index: gtk4-4.14.1+ds/testsuite/reftests/reftest-compare.h
===================================================================
--- gtk4-4.14.1+ds.orig/testsuite/reftests/reftest-compare.h	2024-03-20 21:24:00.036733677 +0100
+++ gtk4-4.14.1+ds/testsuite/reftests/reftest-compare.h	2024-03-20 21:24:00.036733677 +0100
@@ -23,6 +23,10 @@
 
 G_MODULE_EXPORT
 GdkTexture *            reftest_compare_textures        (GdkTexture             *texture1,
-                                                         GdkTexture             *texture2);
+                                                         GdkTexture             *texture2,
+                                                         guint                  *max_diff_out,
+                                                         guint                  *pixels_changed_out,
+                                                         guint                  *pixels_out);
 
 G_END_DECLS
+
Index: gtk4-4.14.1+ds/testsuite/gsk/normalize.c
===================================================================
--- gtk4-4.14.1+ds.orig/testsuite/gsk/normalize.c	2024-03-20 21:24:00.036733677 +0100
+++ gtk4-4.14.1+ds/testsuite/gsk/normalize.c	2024-03-20 21:24:00.036733677 +0100
@@ -9,6 +9,9 @@
   GskRenderer *renderer = gsk_ngl_renderer_new ();
   graphene_rect_t bounds1, bounds2;
   GdkTexture *texture1, *texture2, *diff;
+  guint max_diff = 0;
+  guint pixels_changed = 0;
+  guint pixels = 0;
   GError *error = NULL;
 
   gsk_renderer_realize_for_display (renderer, gdk_display_get_default (), &error);
@@ -20,7 +23,8 @@
   texture1 = gsk_renderer_render_texture (renderer, node1, &bounds1);
   texture2 = gsk_renderer_render_texture (renderer, node2, &bounds2);
 
-  diff = reftest_compare_textures (texture1, texture2);
+  diff = reftest_compare_textures (texture1, texture2,
+                                   &max_diff, &pixels_changed, &pixels);
   g_assert_null (diff);
 
   g_object_unref (texture1);
Index: gtk4-4.14.1+ds/tools/gtk-rendernode-tool-compare.c
===================================================================
--- gtk4-4.14.1+ds.orig/tools/gtk-rendernode-tool-compare.c	2024-03-20 17:33:55.000000000 +0100
+++ gtk4-4.14.1+ds/tools/gtk-rendernode-tool-compare.c	2024-03-20 21:26:45.899590456 +0100
@@ -74,6 +74,9 @@
   GdkTexture *texture[2];
   GdkTexture *diff;
   GError *error = NULL;
+  guint max_diff = 0;
+  guint pixels_changed = 0;
+  guint pixels = 0;
 
   g_set_prgname ("gtk4-rendernode-tool compare");
   context = g_option_context_new (NULL);
@@ -113,8 +116,8 @@
         }
     }
 
-  diff = reftest_compare_textures (texture[0], texture[1]);
-
+  diff = reftest_compare_textures (texture[0], texture[1],
+                                   &max_diff, &pixels_changed, &pixels);
   if (opt_filename && diff)
     {
       if (!gdk_texture_save_to_png (diff, opt_filename))
